### PR TITLE
selfie camera auto-adjustment [FIX]

### LINKED ISSUE #32: Selfie Camera view auto adjustment to the body size
## Service

* [ ] Module Importer
* [ ] Multiplay Component
* [ ] Interaction
* [ ] Gesture
* [x] Screenshot
* [ ] Leaderboard
* [ ] Mannequin
* [ ] Product
* [ ] Localisation
* [ ] QuickChat

## Version
Write the version that you are currently using.
ZEPETO.Wolrd 1.21.2 , Module Importer 1.0.2 , ScreenShot 1.0.1
## Development Environment
Write the version of Unity, OS and so on.
2020.3.9f1,  Mac
## Current Behavior
Write a description of the current operation.
When I change the scale of the ZEPETO character and activate selfie mode, the camera position does not automatically adjust to fit the size of the modified character.
## Expected Behavior
Write a description of the future action.
Please make the selfie camera's position adjust automatically to fit the character's body type. When I change the character's body type, the selfie camera does not show the character's haed.

image

### PR DESCRIPTION
**Pull request by magmukendi**
- Fixed the issue #32 
- Updated the screenshot module to v1.0.2

Here is a demo video
</img>

Close #32

### CODE REVIEW

**Review by bohyunjung — COMMENTED** [review]
@magmukendi

Before reviewing the logic, we must clarify whether the 'height' refers to the character scale or the height associated with ZEPETO Body type.

For now, comments are made for other parts of the request.

File: Project/zepeto-modules/Assets/Zepeto ScreenShot Module/ZepetoScript/ScreenShotModeManager.ts [hunk 1]
```diff
@@ -30,6 +31,7 @@ export default class ScreenShotModeManager extends ZepetoScriptBehaviour {
         ZepetoPlayers.instance.OnAddedLocalPlayer.AddListener(() => {
             this.localPlayer = ZepetoPlayers.instance.LocalPlayer.zepetoPlayer;
             this.zepetoCamera = ZepetoPlayers.instance.LocalPlayer.zepetoCamera.camera;
+            this.myCharacter = Object.FindObjectOfType<ZepetoCharacter>();
```
[the remark this answers is not in the document]
Comment (magmukendi) [reply Project/zepeto-modules/Assets/Zepeto ScreenShot Module/ZepetoScript/ScreenShotModeManager.ts]: Nope, figured out it doesn't.

I made some changes on line 34 [here](https://github.com/JasperGame/zepeto-modules/blob/0c9d30bc1efded1e5acc52b4aabbc18f5a0e6dc7/Project/zepeto-modules/Assets/Zepeto%20ScreenShot%20Module/ZepetoScript/ScreenShotModeManager.ts#L34).

File: Project/zepeto-modules/Assets/Zepeto ScreenShot Module/ZepetoScript/SelfieCamera.ts [hunk 2]
```diff
@@ -125,12 +131,31 @@ export default class SelfieCamera extends ZepetoScriptBehaviour {
...
+            this.distance = this.distance / this.height  * this._myCharacterScaleX 
+            this.height =  this.height * this._myCharacterScaleX; 
+        }
+        else if(this._myCharacterScaleX == 0)
+        {
+            console.log(`Zepeto Character scale CANNOT BE ${this._myCharacterScaleX}`)
+        }
+        else{
+            console.log(`Zepeto character scale CANNOT BE ${this._myCharacterScaleX}`)
+        }
```
[the remark this answers is not in the document]
Comment (magmukendi) [reply Project/zepeto-modules/Assets/Zepeto ScreenShot Module/ZepetoScript/SelfieCamera.ts]: Should have ask myself that question before. 
Will fix that in the next commit.

**Comment by magmukendi** [comment]
So, In this commit [0c9d30b](https://github.com/JasperGame/zepeto-modules/pull/36/commits/0c9d30bc1efded1e5acc52b4aabbc18f5a0e6dc7), I have:
- refactored the variable myCharacter
- removed the redundant if else statement

And in this commit [bb102cc](https://github.com/JasperGame/zepeto-modules/pull/36/commits/bb102cc62a59c928884263353ff4688371e644d3), I have:
- hidden the myCharacter variable from the inspector 
- created a function (`setCameraHeight()`) that calculate the camera height according to the character LeftEye position (using the human body bones). 
- used the camera height and initial distance to compute the new distance between the character and the camera.

**Comment by magmukendi** [comment]
@bohyunjung,

There was something wrong with the AdjustSelfieCamera function in the `SelfieCamera.ts.`. I made some more changes, Here are demo videos of how it is working now,

https://github.com/JasperGame/zepeto-modules/assets/64711820/66d40de8-a880-48cb-b259-2a42bf614c97


https://github.com/JasperGame/zepeto-modules/assets/64711820/92b913be-e8b9-4cef-b093-e1b2410875ea

File: Project/zepeto-modules/Assets/Zepeto ScreenShot Module/ZepetoScript/SelfieCamera.ts [hunk 3]
```diff
@@ -125,12 +129,38 @@ export default class SelfieCamera extends ZepetoScriptBehaviour {
...
+            this.distance = this.distance / this.height;
+        }
+        else
+        {
+            this.distance = this.distance / (this.height * initialHeight);
+        }
+    }
+
+    // This function returns the camera height following the head position;
+    private getSelfieCameraHeight(localPlayer:ZepetoPlayer): number {
```
Comment (bohyunjung) [inline Project/zepeto-modules/Assets/Zepeto ScreenShot Module/ZepetoScript/SelfieCamera.ts:158]: What do you think of clarifying that this returns a 'new', 'processed' value for the selfie camera height?

I suggest `calculateSelfieCameraHeight`.

**Review by bohyunjung — COMMENTED** [review]
@magmukendi Looks like a good solution for selfie height adjustment.
I've posted my opinion on a method. Please check.

File: Project/zepeto-modules/Assets/Zepeto ScreenShot Module/ZepetoScript/SelfieCamera.ts (hunk 3 above)
Comment (magmukendi) [reply Project/zepeto-modules/Assets/Zepeto ScreenShot Module/ZepetoScript/SelfieCamera.ts:158]: Oh yeah, this makes more sense. I'll update this as soon as I can.

File: Project/zepeto-modules/Assets/Zepeto ScreenShot Module/ZepetoScript/SelfieCamera.ts [hunk 4]
```diff
@@ -1,8 +1,10 @@
 import { ZepetoScriptBehaviour } from 'ZEPETO.Script'
-import { Vector3, Transform, Mathf, Time, Quaternion, HideFlags, GameObject, Input, Application } from 'UnityEngine'
+import { Vector3, Transform, Mathf, Object, Time, Quaternion, HideFlags, GameObject, Input, Application, HumanBodyBones } from 'UnityEngine';
 import { EventSystem } from 'UnityEngine.EventSystems';
+import ScreenShotModeManager from './ScreenShotModeManager';
+~import { ZepetoPlayer } from 'ZEPETO.Character.Controller';
```
Comment (bohyunjung) [inline Project/zepeto-modules/Assets/Zepeto ScreenShot Module/ZepetoScript/SelfieCamera.ts:5]:
```suggestion
import { ZepetoPlayer } from 'ZEPETO.Character.Controller';
```

**Review by bohyunjung — APPROVED** [review]
👍

**Pull request merged**